From ff8a6f5e84889979dc2de7f6a437654b9a6ac593 Mon Sep 17 00:00:00 2001
From: Matthew Gerlach <matthew.gerlach@linux.intel.com>
Date: Tue, 8 Jun 2021 08:30:46 -0700
Subject: [PATCH 22/93] fpga: dfl: fix VF creation when ports have no local BAR
 space

When a port is not connected to the same PCIe endpoint as
the FME, the port does not need to be released before being
virtualized.  Fix VF creation code to handle this new use
case.

Signed-off-by: Matthew Gerlach <matthew.gerlach@linux.intel.com>
---
 drivers/fpga/dfl.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/fpga/dfl.c b/drivers/fpga/dfl.c
index 2d2140ce0cd0..b1b358b40f39 100644
--- a/drivers/fpga/dfl.c
+++ b/drivers/fpga/dfl.c
@@ -1751,15 +1751,22 @@ EXPORT_SYMBOL_GPL(dfl_fpga_cdev_config_ports_pf);
 int dfl_fpga_cdev_config_ports_vf(struct dfl_fpga_cdev *cdev, int num_vfs)
 {
 	struct dfl_feature_dev_data *fdata;
-	int ret = 0;
+	int ret = 0, port_count = 0;
 
 	mutex_lock(&cdev->lock);
+
+	list_for_each_entry(fdata, &cdev->port_dev_list, node) {
+		if (fdata->dev)
+			continue;
+		port_count++;
+	}
+
 	/*
 	 * can't turn multiple ports into 1 VF device, only 1 port for 1 VF
 	 * device, so if released port number doesn't match VF device number,
 	 * then reject the request with -EINVAL error code.
 	 */
-	if (cdev->released_port_num != num_vfs) {
+	if (port_count && (cdev->released_port_num != num_vfs)) {
 		ret = -EINVAL;
 		goto done;
 	}
-- 
2.39.0

